metric/metric_recorder.go

```diff
@@ -0,0 +1,44 @@
+package metric
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	"github.com/fabric8-services/fabric8-wit/log"
+	"github.com/goadesign/goa"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var reqLabels = []string{"entity", "action"}
+
+// Recorder record metrics
+func Recorder() goa.Middleware {
+	reqCnt := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "wit_requests_total",
+		Help: "Total number of WIT requests.",
+	}, reqLabels)
+	// TODO VN: Check err, need prometheus version change.
+	prometheus.Register(reqCnt)
+
+	return func(h goa.Handler) goa.Handler {
+		return func(ctx context.Context, rw http.ResponseWriter, req *http.Request) error {
```

Comment:
The sample output for metric when invoking http://localhost:8080/metrics
**Output:**
# HELP wit_requests_total Total number of WIT requests.
# TYPE wit_requests_total counter
wit_requests_total{action="create",entity="Space"} 2
wit_requests_total{action="list",entity="Space"} 2
wit_requests_total{action="list",entity="Users"} 1
wit_requests_total{action="list",entity="Workitem"} 1
wit_requests_total{action="show",entity="Space"} 1

This handler is called as last handler as expected but there are some exception which listed here:

1. If you make this call: 
./bin/wit-cli list space-areas --host 127.0.0.1:8080 --key $token --pp

Here, mandatory spaceID is missing so call will fail but metril_recorder is called with entity=Space and action=show.
It means, whenever **list** call made for any child entity without specifying parent at that time it will implicitly call parent with **show** action.  This may be due to some design flaw or bug and may required different investigation.

Reply:
This is due to the order of middlewares. Recorder is before url validation. 

But I think this is fine. I think Total should be Total regardless of state of the request. The state of the request has different counters.

Reply:
Please note, we are making **'list'** call for **'X'** entity (missing required param) and metric is recorded for **'Y'** entity with action **'show'** and NOT **'list'**.  So both entity and action are different here.

Reply:
Ah yes. Sorry, miss read. Known issue, but unrelated to metrics.